File: cs315_codes/ch5_scope_examples/PHP/nested_functions.php
<!--
   The scope of a variable is the context within which it is defined. 
   For the most part all PHP variables only have a single scope.
   It is possible access the global variables 
-->
<?php
  $g = 3;     // a global variable, defined outside of all functions

  function foo() {

     $x = 7;
     function bar() {  // nested functions can be defined, but

        echo "x=$x <br>"; // the nonlocal variables are not accessible.

        global $g;
        echo "g=$g <br>"; // the global variables are accessible.
     }
     bar();
  } // foo

foo();
?>

File: cs315_codes/ch5_scope_examples/C/nested-functions.c
/*
Nested function is not supported by C 
because we cannot define a function within another function in C. 

Nested functions are supported as an extension in GNU C, 
but are not supported by GNU C++.
*/

#include<stdio.h>
void sub1() {
  int x;
  void sub2() {
    int y;
    void sub3() {
      int z = 3;
      printf("In sub3, x=%d, y=%d, z=%d\n", x, y, z);
    } // sub3
    y = 2;
    sub3();
    printf("In sub2, x=%d, y=%d\n", x, y);
  } // sub2
  x = 1;
  sub2();
  printf("In sub1, x=%d\n", x);
} // sub1

void main() {
  sub1();  
}


